Add guide to next vercel logflare logging (#1926)
vercel/docs: pages/guides/ultimate-guide-to-logging-nextjs-vercel-logflare.mdx

<Link
  href="https://vercel.com/integrations/logflare"
  underlineOnHover={false}
  target="_blank"
>
  <Button primary>Install Logflare Vercel Integration</Button>
</Link>

### Step 2: Setup a Log Drain

Create a Log Drain for your Vercel project.

<Image
  src={`${
    process.env.ASSETS
  }/guides/ultimate-guide-to-logging-nextjs-vercel-logflare/create-vercel-log-drain.png`}
  width={1608 / 2}
  height={1199 / 2}
  oversize
/>

<Note>
  You'll want to select a project for your new Log Drain. Otherwise all logs for
  all projects will go into the same Logflare source. While you can do this,
  it's recommended to send logs from each project to a different Logflare
  source.
</Note>

### Step 3: Verify Logs

Immediately after you set up a Log Drain, you should see logs streaming into your Logflare account.

<Link
  href="https://logflare.app/dashboard"
  underlineOnHover={false}
  target="_blank"
>
  <Button primary>Visit Your Logflare Dashboard</Button>
</Link>

## Using Structured Log Events

### Lambda Reports

> Logflare is different from other logging products in that we specifically designed an ingest endpoint for Vercel. The `lambdaMessage` that comes over is the console output from the lambda that runs on Verce's infrastructure. This comes over as a string. Logflare automatically parses it into the `parsedLambdaMessage` object for you.

Reviewer: ```suggestion
Logflare is different from other logging products in that we specifically designed an ingest endpoint for Vercel. The `lambdaMessage` that comes over is the console output from the lambda that runs on Vercel's infrastructure. This comes over as a string. Logflare automatically parses it into the `parsedLambdaMessage` object for you.
```

              "x_vercel_trace": "sfo1"
            },
            "method": "GET",
            "url": "/api/date"
          },
          "response": {
            "statusCode": 200
          },
          "time": 1592333966191,
          "v": 1
        }
      }
    ],
    "parse_status": "full",
    "report": {
      "billed_duration_ms": 100,
      "duration_ms": 11,
      "max_memory_used_mb": 76,
      "memory_size_mb": 1024
    },
    "request_id": "40d796c7-b1fa-45da-93dc-d0a6c9da9285"
  },
  "path": "api/date",
  "projectId": "QmRxhyrspZA1hKp4ft1yQCGeq6nVyyvyd2mv2fhyWEeEdw",
  "proxy": {
    "clientIp": "68.6.69.18",
    "host": "pino-logflare-nextjs-vercel-example.now.sh",
    "method": "GET",
    "path": "/api/date",
    "referer": "https://pino-logflare-nextjs-vercel-example.now.sh/",
    "region": "sfo1",
    "scheme": "https",
    "statusCode": 200,
    "timestamp": 1592333966164,
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36"
  },
  "requestId": "f2x8s-1592333966164-93acc6b86c61",
  "source": "lambda",
  "statusCode": 200,
  "timestamp": 1592333966170
}
```

> ## Deploy the Example Repo
> 
> 🎉 Let's get started! Deploy the example repo.
> 
> <DeployButton url="https://vercel.com/new/project?template=https://github.com/Logflare/pino-logflare-nextjs-vercel-example" />
> 
> ## The Example Repo in Detail
> 
> What exactly have we done in our example repo?
> 
> ### Logging the Github Commit
> 
> Set the fields to be included with every log event in the `base` object when you instantiate Pino.
> 
> Logging the Github commit with every event is useful for tying defects to a specific deploy!
> 
> ```jsx
> // logger.js
> 
> import pino from 'pino';
> 
> export default pino(
>     {
>         level: 'debug',
>         base: {
>             env: process.env.ENV || "ENV not set",
>             revision: process.env.VERCEL_GITHUB_COMMIT_SHA,
> 
>     }
> );
> ```

Reviewer 2: ```suggestion
## Example Repository

Let’s walk through the example repository, which is available on GitHub: [pino-logflare-nextjs-vercel-example](https://github.com/Logflare/pino-logflare-nextjs-vercel-example).

You can deploy it on Vercel by clicking the button below:

<DeployButton url="https://vercel.com/new/project?template=https://github.com/Logflare/pino-logflare-nextjs-vercel-example" />

### Logging the Github Commit

We created a logger ([logger.js](https://github.com/Logflare/pino-logflare-nextjs-vercel-example/blob/main/logger/logger.js)) which sets the fields to be included with every log event in the `base` object when you instantiate Pino.

Logging the Github commit with every event is useful for tying defects to a specific deploy.

```jsx
// logger/logger.js
import pino from 'pino';

export default pino(
  {
    level: 'debug',
    base: {
      env: process.env.ENV || "ENV not set",
      revision: process.env.VERCEL_GITHUB_COMMIT_SHA,
    }
  }
);
```
```

### Logging Request Headers

> Let's log all our request headers so we can understand which user agents are making requests to our service. `getServerSideProps` gets called whenever a request is made to our server. It's called with a `context` which has a bunch of interesting data about the request we can introspect with structured logging!`getServerSideProps` gets called whenever a request is made to our server. It's called with a `context` which has a bunch of interesting data about the request we can introspect with structured logging!
> 
> To get the request headers we just need to access them via `getServerSideProps`:

Reviewer 2: ```suggestion
Let's log all our request headers so we can understand which user agents are making requests to our service. `getServerSideProps` gets called whenever a request is made to our server. It's called with a `context` which has a bunch of interesting data about the request we can introspect with structured logging.
```

```jsx

> // index.js
> 
> import { prepObjectKeys } from "./utils.js"
> 
> const logger = require('pino') from "./logger.js"

Reviewer 2: ```suggestion
// pages/index.js

import { prepObjectKeys } from "../logger/utils.js"
import logger from "../logger/logger.js"

// ...
```

Author: Ah yep sorry old stuff!  Thanks!!

export async function getServerSideProps(context) {

  const headers = prepObjectKeys(context.req.headers)

  logger.info({ request: { headers: headers } }, "Index request")

  return {
    props: {}, // will be passed to the page component as props
  }
}
```

> <Note>
>   Notice that we need to modify our request header keys before sending them to
>   Logflare. Logflare uses BigQuery directly as it's storage engine. Logflare
>   turns each field in an event into it's own BigQuery column, and BigQuery has
>   limitation around column names. Specifically, a column name must contain only
>   letters (a-z, A-Z), numbers (0-9), or underscores (_), and it must start with
>   a letter or underscore.
> </Note>
> 
> To make our request header object compatible with Logflare and BigQuery lets replace the dashes with underscores:

Reviewer 2: I don’t think this needs to be a `Note` component because the next paragraph also talks about the same thing.

```suggestion
Notice that we need to modify our request header keys before sending them to Logflare. Logflare uses BigQuery directly as it's storage engine. Logflare turns each field in an event into it's own BigQuery column, and BigQuery has limitation around column names. Specifically, a column name must contain only letters (a-z, A-Z), numbers (0-9), or underscores (_), and it must start with a letter or underscore.

To make our request header object compatible with Logflare and BigQuery, we call `prepObjectKeys` which replaces the dashes with underscores:
```

```jsx

> // utils.js

Reviewer 2: ```suggestion
// logger/utils.js
```

const prepObjectKeys = headers => {
  const keyValues = {}
  Object.keys(headers).map(key => {
    const newKey = key.replace(/-/g, '_')
    keyValues[newKey] = headers[key]
  })

  return keyValues
}

export { prepObjectKeys }
```

## Setup the Logflare Vercel Integration with Your Project

> Now that we're logging some useful data lets setup the Logflare Vercel integration to get our newly structured logs going to Logflare.

Reviewer 2: ```suggestion
Now that we're logging some useful data, let’s set up the Logflare Vercel integration to get our newly structured logs going to Logflare.
```

### Step 1: Install the Integration

> During the install process you will sign up for a Logflare account (or sign in if you already have one) and setup a new Logflare source. Logflare `source`s are where you logs will go.

Reviewer 2: ```suggestion
Click the button below to install the integration. During the install process, you will sign up for a Logflare account (or sign in if you already have one) and set up a new Logflare source. Logflare `source`s are where your logs will go.
```

<Link
  href="https://vercel.com/integrations/logflare"
  underlineOnHover={false}
  target="_blank"
>
  <Button primary>Install Logflare Vercel Integration</Button>
</Link>

### Step 2: Setup a Log Drain

Create a Log Drain for your Vercel project.

<Image
  src={`${
    process.env.ASSETS
  }/guides/ultimate-guide-to-logging-nextjs-vercel-logflare/create-vercel-log-drain.png`}
  width={1608 / 2}
  height={1199 / 2}
  oversize
/>

<Note>

>   You'll want to select a project for your new Log Drain otherwise all logs for

Reviewer 2: ```suggestion
  You'll want to select a project for your new Log Drain. Otherwise all logs for
```

  all projects will go into the same Logflare source. While you can do this,
  it's recommended to send logs from each project to a different Logflare
  source.
</Note>

### Step 3: Verify Logs

> Immediately after you setup a Log Drain you should see logs streaming into your Logflare account.

Reviewer 2: ```suggestion
Immediately after you set up a Log Drain, you should see logs streaming into your Logflare account.
```

<Link
  href="https:logflare.app/dashboard"
  underlineOnHover={false}
  target="_blank"
>
  <Button primary>Visit Your Logflare Dashboard</Button>
</Link>

## Using Structured Log Events

### Lambda Reports

> Logflare is different from other logging products in that we specifically designed an ingest endpoint for Vercel. The `lambdaMessage` that comes over is the console output from the lambda that runs on Verce's infrastructure. This comes over as a string. There is some really great data in there so Logflare automatically parses it into the `parsedLambdaMessage` object for you.

Reviewer 2: ```suggestion
Logflare is different from other logging products in that we specifically designed an ingest endpoint for Vercel. The `lambdaMessage` that comes over is the console output from the lambda that runs on Verce's infrastructure. This comes over as a string. Logflare automatically parses it into the `parsedLambdaMessage` object for you.
```

The `report` object inside the `parsedLambdaMessage` is interesting. It reports `billing_duration_ms`, `duration_ms`, `max_memory_used_mb`, and `memory_size_mb`.

<Note>
  Easily chart the average response time of our lambdas in Logflare with the

>   query `c:avg(m.parsedLambdaMessage.report.duration_ms) c:group_by(t::minute)`

Reviewer 2: ```suggestion
  query `c:avg(m.parsedLambdaMessage.report.duration_ms) c:group_by(t::minute)`.
```

</Note>

### Example Searches for Vercel Infrastructure Logs

There are lots of useful fields in the infrastructure logs provided by Vercel.

All fields of the logged event are accessible with the Logflare Query Language under the `metadata` (or `m` ) object.

<Image
  src={`${
    process.env.ASSETS
  }/guides/ultimate-guide-to-logging-nextjs-vercel-logflare/vercel-lambda-search-logflare.png`}
  width={1608 / 2}
  height={890 / 2}
  oversize
/>

> `m.source:"static"` - Assets served from the Vercel CDN
> 
> `m.source:"lambda"` - Function calls
> 
> `m.source:"build"` - Build logs
> 
> `m.proxy.statusCode:>499` - Requests with HTTP status code greater than 499
> 
> `m.proxy.userAgent:~"bot"` - All user agents with `bot` in the string
> 
> `m.path:~"signup/success"` - See all signups (assuming they get sent to a url containing the string `signup/success`

Reviewer 2: ```suggestion
- `m.source:"static"` - Assets served from the Vercel CDN
- `m.source:"lambda"` - Function calls
- `m.source:"build"` - Build logs
- `m.proxy.statusCode:>499` - Requests with HTTP status code greater than 499
- `m.proxy.userAgent:~"bot"` - All user agents with `bot` in the string
- `m.path:~"signup/success"` - See all signups (assuming they get sent to a url containing the string `signup/success`
```

### Example Data Studio Dashboard

<Video
  src={`${
    process.env.ASSETS
  }/guides/ultimate-guide-to-logging-nextjs-vercel-logflare/explore-vercel-logs.mp4`}
  width={1440 / 1.5}
  height={900 / 1.5}
  oversize
  controls
  muted
  autoPlay
  loop
/>

> Logflare integrates seemlessly with Google Data Studio. The Logflare Vercel integration guide has an [example Data Studio dashboard](https://logflare.app/guides/vercel-setup#example-dashboard) you can easily copy once you have a data source setup in Data Studio.

Reviewer 2: ```suggestion
Logflare integrates seamlessly with Google Data Studio. The Logflare Vercel integration guide has an [example Data Studio dashboard](https://logflare.app/guides/vercel-setup#example-dashboard) you can easily copy once you have a data source set up in Data Studio.
```

### Example Alerts

> With Logflare you can use the same Logflare query to define alerts as you use to query. Lets setup an SMS alert so we get a text message whenever Vercel finishes deploying our changes.

Reviewer 2: ```suggestion
With Logflare you can use the same Logflare query to define alerts as you use to query. Let’s set up an SMS alert so we get a text message whenever Vercel finishes deploying our changes.
```

> First add a new source to send the log event associated with a finished build. We'll call it `pino-logflare-vercel.build-done`. Edit the source to enable SMS alerts:

Reviewer 2: ```suggestion
First, add a new source to send the log event associated with a finished build. We'll call it `pino-logflare-vercel.build-done`. Edit the source to enable SMS alerts:
```

<Image
  src={`${
    process.env.ASSETS
  }/guides/ultimate-guide-to-logging-nextjs-vercel-logflare/setup-logflare-alert-vercel-build.png`}
  width={1608 / 2}
  height={1051 / 2}
  oversize
/>

And now we simply send our build done message to that source with a rule on our main ingest source.

<Image
  src={`${
    process.env.ASSETS
  }/guides/ultimate-guide-to-logging-nextjs-vercel-logflare/logflare-vercel-event-routing-rules.png`}
  width={1608 / 2}
  height={1057 / 2}
  oversize
/>

## Why BigQuery for Logs and Metrics?

BigQuery is an amazing tool for storing and querying lots of data. It was [originally designed](https://cloud.google.com/files/BigQueryTechnicalWP.pdf) specifically to store and query click data from Google's ad network. They simply wanted a time series chart. With BigQuery we can quickly derive time series data from billions of structured events. And then continue to derive these metrics dynamically with SQL. We don't need to define new "metrics" for each version of a "metric" we're interested in.

> Lets say we have users on paid plans. Typically with StatsD type "metrics" we'd have to define a new metric for user logins for each plan. This doesn't seem terrible, but what if we had 100 attributes per login? We'd need a separate "metric" for every combination of attribute someone would potentially want to know about. And we'd have to set it all up in advance.

Reviewer 2: ```suggestion
Let’s say we have users on paid plans. Typically with StatsD type "metrics" we'd have to define a new metric for user logins for each plan. This doesn't seem terrible, but what if we had 100 attributes per login? We'd need a separate "metric" for every combination of attributes someone would potentially want to know about. And we'd have to set it all up in advance.
```

Instead of doing that, we can just log the structured event with whatever data we want and retroactively query it in any unknown combination we may think of in the future.

With Logflare and BigQuery we just send a structured event each time a user logs in, with fields about that user. So we can just query for users by plan. The Logflare Query Language query would simply be `m.user_login.plan:"pro"`.

And now, since we have all the individual events backing our metrics, we can drill into the specific events making up each point in time in our charts. So if there is a spike in user logins, we can drill into that spike and see exactly who logged in when and how many times. With typical "metrics" like from StatsD this is impossible.

> ## Bonus: Send Core Web Vitals to Logflare
> 
> Next, setup your Next.js app to send Core Web Vitals to Logflare!
> 
> \#TODO Link to Core Web Vitals Logflare guide here.

Reviewer 2: Please finish this or delete it:

```suggestion
```

Author: I have this guide in the hopper here but I'll delete it for now.
